js/ag_reference.js

```diff
@@ -905,3 +921,181 @@ AttributeGroupReferenceAggregateFn.prototype = {
+ *  - Clients _do not_ directly access this constructor.
+ *  - This will currently be used by the aggregateGroup histogram function to return a
+ *    BucketAttributeGroupReference rather than a {@link ERMrest.Reference}
+ *
+ * @param       {ERMRest.AttributeGroupColumn[]} keyColumns List of columns that will be used as keys for the attributegroup request.
+ * @param       {?ERMRest.AttributeGroupColumn[]} aggregateColumns List of columns that will create the aggreagte columns list in the request.
+ * @param       {ERMRest.AttributeGroupLocation} location  The location object.
+ * @param       {ERMRest.Catalog} catalog  The catalog object.
+ * @param       {Object} options contains bucketCount, min, max
+ * @constructor
+ */
+function BucketAttributeGroupReference(keyColumns, aggregateColumns, location, catalog, options) {
+    // call the parent constructor
+    BucketAttributeGroupReference.superClass.call(this, keyColumns, aggregateColumns, location, catalog);
+
+    this._options = options;
+}
+
+// extend the prototype
+module._extends(BucketAttributeGroupReference, AttributeGroupReference);
+
+// properties to be overriden:
+BucketAttributeGroupReference.prototype.sort = function (sort) {
+    verify(true == false, "Invalid function");
+};
+
+BucketAttributeGroupReference.prototype.search = function (term) {
+    verify(true == false, "Invalid function");
+};
+
+/**
+ * @return {ERMRest.BucketAttributeGroupPage}
+ */
+BucketAttributeGroupReference.prototype.read = function () {
+    var moment = module._moment;
+
+    // function decimals_s(num){
+    //     var num_s = "" + num;
+    //     return num_s.split(".")[1].length;
+    // }
+    //
+    // function addFloats(min, width) {
+    //     var precision = Math.max(decimals_s(min), decimals_s(width));
+    //     return ((min*1) + (width*1)).toFixed(precision);
+    // }
+
+    function calculateWidthLabel(min, binWidth) {
+        var nextLabel;
+        if (currRef._keyColumns[0].type.rootName.indexOf("date") > -1)  {
+            nextLabel = moment(min).add(binWidth, 'd').format(module._dataFormats.DATE);
+        } else if (currRef._keyColumns[0].type.rootName.indexOf("timestamp") > -1) {
+            nextLabel = moment(min).add(binWidth, 's').format(module._dataFormats.DATETIME.display);
+        } else {
+            nextLabel = (min + binWidth);
+        }
+        return nextLabel;
+    }
+
+    try {
+        var defer = module._q.defer();
+
+        var uri = this.uri;
+
+        var currRef = this;
+        this._server._http.get(uri).then(function (response) {
+            var data = {
+                x: [],
+                y: []
+            };
+
+            var labels = {
+                min: [],
+                max: []
+            };
+
+            var min, max;
+
+            for (var i=0; i<response.data.length; i++) {
+                var index = response.data[i].c1[0];
+                if (index !== null) {
+                    min = response.data[i].c1[1];
```

Comment:
You're assuming c1 and c2 are the name of columns here, but in the reference constructor your accepting any list of columns. Since this is completely written for this specific binning api in ermrest, you can change the Reference constructor variables to be the actual column name that you want binning for, then constructor the column object in the constructor.